### PR TITLE
5 pull request

### PR DESCRIPTION
**Pull request by sawkiewicz-tomasz**

### CODE REVIEW

File: src/main/java/sawkiewicz/task1/CartItem.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package sawkiewicz.task1;
+
+public class CartItem {
+    int quantity;
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/CartItem.java]: fields always should be `private`, rule of thumb is to keep class internals for it's own use

File: src/main/java/sawkiewicz/task1/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        this.quantity = quantity;
+        this.item = item;
+    }
+
+    public CartItem() {
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/CartItem.java]: `equals` has a special meaning, the same as `hashCode` https://www.programcreek.com/2011/07/java-equals-and-hashcode-contract/

so, please don't use it here!

if you need some type of comparison implement `isSameAs` or in your case `hasSameQuantity`

File: src/main/java/sawkiewicz/task1/Item.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public Item(String id, int price, int threshold, int specialPrice) {
+        this.id = id;
+        this.price = price;
+        this.threshold = threshold;
+        this.specialPrice = specialPrice;
+    }
+
+    public void setId(String id) {
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/Item.java]: why do you expose all those setters?

File: src/main/java/sawkiewicz/task1/Item.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    public int getSpecialPrice() {
+        return specialPrice;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/Item.java]: the same as above, read about equals and hash code contract

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package sawkiewicz.task1;
+
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+
+
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: empty line - please remove

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+@RestController
+public class ItemController {
+    ItemRepository itemRepository = new ItemRepository();
+    ShoppingCartRepository shoppingCartRepository = new ShoppingCartRepository();
+
+
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: again , empty line https://www.romexsoft.com/blog/improve-java-code-quality/

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    @RequestMapping(method = RequestMethod.GET, value = "/total/{cartId}")
+    ResponseEntity<String> showTotalPrice(@PathVariable String cartId) {
+        return ResponseEntity.ok("Total price is: " + shoppingCartRepository.findById(cartId).calculateTotalCost());
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: according to Rest specification, you should return an object like the one below

```
{
     "response": "Ok"
     "message": ""
}
```

Comment (marcindziedzic) [reply src/main/java/sawkiewicz/task1/ItemController.java]: Please read about JSON https://www.copterlabs.com/json-what-it-is-how-it-works-how-to-use-it/

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @RequestMapping(method = RequestMethod.POST, value = "/cart/{cartId}/item")
+    public ResponseEntity<String> addItem(@PathVariable String cartId, @RequestBody Item itemToAdd) {
+        CartItem cartItemToAdd = new CartItem(1, itemToAdd);
+
+        int sizeOfPurchase = shoppingCartRepository.findById(cartId).purchase.size();
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: please split into multiple lines

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        System.out.println(sizeOfPurchase);
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: don't use `println`, that's a bad practise

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+        if (sizeOfPurchase == 0) {
+            shoppingCartRepository.findById(cartId).purchase.add(cartItemToAdd);
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: again, multiple lines

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        } else if (shoppingCartRepository.findById(cartId).purchase.contains(cartItemToAdd)) {
+            for (CartItem cartItem : shoppingCartRepository.findById(cartId).purchase) {
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: multiple lines

File: src/main/java/sawkiewicz/task1/ItemController.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                if (cartItem.item.getId().equals(itemToAdd.getId())) {
+                    int temp = cartItem.getQuantity();
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemController.java]: maybe instead 

`cartItem.increaseQuantity(...)`

File: src/main/java/sawkiewicz/task1/ItemRepository.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+    static {
+        items.add(new Item("A", 40, 3, 70));
+        items.add(new Item("B", 10, 2, 15));
+        items.add(new Item("C", 30, 4, 60));
+        items.add(new Item("D", 25, 2, 40));
+    }
+
+    public static List<Item> getItems() {
+
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemRepository.java]: empty line, remove

File: src/main/java/sawkiewicz/task1/ItemRepository.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        return ImmutableList.copyOf(items);
+    }
+
+    @Override
+    public Item findById(String id) {
+
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemRepository.java]: empty line , remove https://softwareengineering.stackexchange.com/questions/17305/how-do-you-use-blank-lines-in-your-code

File: src/main/java/sawkiewicz/task1/ItemRepository.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        return items.stream().filter(item -> item.getId().equals(id))
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ItemRepository.java]: move `.filter` to new line

File: src/main/java/sawkiewicz/task1/Repository.java [hunk 16]
```diff
@@ -0,0 +1,6 @@
+package sawkiewicz.task1;
+
+public interface Repository {
+
+    Object findById(String id);
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/Repository.java:5]: please , use generics

File: src/main/java/sawkiewicz/task1/ShoppingCart.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
+package sawkiewicz.task1;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ShoppingCart {
+    String id;
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ShoppingCart.java]: fields should be private

File: src/main/java/sawkiewicz/task1/ShoppingCart.java [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    List <CartItem> purchase = new ArrayList<>();
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ShoppingCart.java]: rename `purchase` to `items`

File: src/main/java/sawkiewicz/task1/ShoppingCart.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    public String getId() {
+        return id;
+    }
+
+    public int calculateTotalCost() {
+        int totalCost = 0;
+
+        for (CartItem cartItem: purchase) {
+
+            int quantity = cartItem.quantity;
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ShoppingCart.java]: if needed, use getters `cartItem.getQuantity()`, fields should be private

File: src/main/java/sawkiewicz/task1/ShoppingCartRepository.java [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+    public List<ShoppingCart> shoppingCarts = new ArrayList<>();
+
+
+    public void add(ShoppingCart cartToAdd){
+        shoppingCarts.add(cartToAdd);
+    }
+
+
+    public List<ShoppingCart> getShoppingCarts(){
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ShoppingCartRepository.java]: is this methods used?

File: src/main/java/sawkiewicz/task1/ShoppingCartRepository.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        return shoppingCarts;
+    }
+
+    @Override
+    public ShoppingCart findById(String id) {
+
+        return shoppingCarts.stream().filter(shoppingCart -> shoppingCart.getId().equals(id))
+                .collect(toCollection(() -> new ArrayBlockingQueue<ShoppingCart>(1)))
```
Comment (marcindziedzic) [inline src/main/java/sawkiewicz/task1/ShoppingCartRepository.java]: `ArrayBlockingQueue` - no way! Instead `ArrayList`

File: src/main/java/sawkiewicz/task1/CartItem.java (hunk 1 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/CartItem.java]: done

File: src/main/java/sawkiewicz/task1/Item.java (hunk 3 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/Item.java]: removed

File: src/main/java/sawkiewicz/task1/ItemController.java (hunk 5 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ItemController.java]: done

File: src/main/java/sawkiewicz/task1/ItemController.java (hunk 9 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ItemController.java]: removed

File: src/main/java/sawkiewicz/task1/ItemController.java (hunk 10 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ItemController.java]: removed

File: src/main/java/sawkiewicz/task1/ItemController.java (hunk 12 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ItemController.java]: done

File: src/main/java/sawkiewicz/task1/ItemRepository.java (hunk 13 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ItemRepository.java]: removed

File: src/main/java/sawkiewicz/task1/ItemRepository.java (hunk 15 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ItemRepository.java]: done

File: src/main/java/sawkiewicz/task1/ShoppingCart.java (hunk 17 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ShoppingCart.java]: done

File: src/main/java/sawkiewicz/task1/ShoppingCart.java (hunk 18 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ShoppingCart.java]: renamed

File: src/main/java/sawkiewicz/task1/ShoppingCart.java (hunk 19 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ShoppingCart.java]: done

File: src/main/java/sawkiewicz/task1/ShoppingCartRepository.java (hunk 20 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ShoppingCartRepository.java]: removed

File: src/main/java/sawkiewicz/task1/ShoppingCartRepository.java (hunk 21 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/ShoppingCartRepository.java]: done

File: src/main/java/sawkiewicz/task1/Item.java (hunk 4 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/Item.java]: done

File: src/main/java/sawkiewicz/task1/CartItem.java (hunk 2 above)
Comment (sawkiewicz-tomasz) [reply src/main/java/sawkiewicz/task1/CartItem.java]: done

**Pull request closed without merging**